File: datasets/generate_RD1d_data.jl
# Script to solve the Reaction-Diffusion equation numerically to construct PDE
# solution datasets for training and sampling.

using HDF5
using Random

# ── Global parameters ─────────────────────────────────────────────────────────

const rho_RD = 0.01
const nu_RD  = 0.005
const Nx_RD  = 128
const xL_RD, xR_RD = 0.0, 1.0
const dt_save = 0.01
const ini_time = 0.0
const fin_time = 1.0 - dt_save

const dx_RD = (xR_RD - xL_RD) / Nx_RD
const x_RD  = collect(xL_RD + 0.5*dx_RD : dx_RD : xR_RD - 0.5*dx_RD)  # cell-centred grid
const nt_RD = Int(ceil((fin_time - ini_time) / dt_save)) + 1
const t_grid_RD = collect(range(ini_time, fin_time, length=nt_RD))

# ── Solver ────────────────────────────────────────────────────────────────────

"""
    courant_diff(dx; epsilon=1e-3)

Diffusive Courant time-step limit.
"""
courant_diff(dx; epsilon=1e-3) = 0.5 * dx^2 / (epsilon + 1e-8)

"""
    flux_rd(u, gL, gR, dx, N)

Compute the diffusive flux for the reaction-diffusion solver.
"""
function flux_rd(u, gL, gR, dx, N)
    u_ext = zeros(N + 4)
    u_ext[3:N+2] .= u
    f = -nu_RD .* (u_ext[3:N+3] .- u_ext[2:N+2]) ./ dx
    f[1] = gL
    f[end] = gR
    return f
end

"""
    update_rd(u, u_tmp, dt, gL, gR, dx)

One Strang-split half-step: reaction (exact) then diffusion (explicit Euler).
"""
function update_rd(u, u_tmp, dt, gL, gR, dx)
    N = length(u)
    stiff = 1.0 ./ (1.0 .+ exp.(-rho_RD .* dt) .* (1.0 .- u) ./ (u .+ 1e-12))
    f = flux_rd(u_tmp, gL, gR, dx, N)
    return stiff .- dt .* (f[2:N+1] .- f[1:N]) ./ dx
end

"""
    solve_single(u0, gL, gR)

Time-integrate the 1D reaction-diffusion equation with initial condition `u0`
and Neumann-type boundary fluxes `gL` (left) and `gR` (right).

Returns the solution array of shape (Nx, nt).
"""
function solve_single(u0, gL, gR)
    u = copy(u0)
    sol = zeros(Float64, Nx_RD, nt_RD)
    sol[:, 1] .= u
    t = ini_time
    save_idx = 2
    CFL = 0.25

    while t < fin_time && save_idx <= nt_RD
        dt = min(courant_diff(dx_RD; epsilon=nu_RD) * CFL,
                 fin_time - t,
                 t_grid_RD[save_idx] - t)
        dt <= 1e-8 && break
        u_tmp = update_rd(u, u, 0.5 * dt, gL, gR, dx_RD)
        u     = update_rd(u, u_tmp, dt, gL, gR, dx_RD)
        t += dt
        if t >= t_grid_RD[save_idx] - 1e-12
            sol[:, save_idx] .= u
            save_idx += 1
        end
    end
    sol[:, end] .= u
    return sol  # (Nx, nt)
end

# ── Initial condition generator ───────────────────────────────────────────────

"""
    generate_ic(xc; k_tot=3, num_choice_k=2)

Generate a random smooth initial condition on the grid `xc` as a sum of sinusoids,
normalised to [0, 1].
"""
function generate_ic(xc; k_tot=3, num_choice_k=2)
    selected = rand(1:k_tot, num_choice_k)
    onehot = zeros(Int, k_tot)
    for j in selected
        onehot[j] += 1
    end
    kk = 2π .* (1:k_tot) .* onehot ./ (xc[end] - xc[1])
    amp = rand(k_tot, 1)
    phs = 2π .* rand(k_tot, 1)
    u = vec(sum(amp .* sin.(kk .* xc' .+ phs), dims=1))

    if rand() < 0.1
        u = abs.(u)
    end
    u .*= rand([-1, 1])

    if rand() < 0.1
        xL_m = rand() * 0.35 + 0.1
        xR_m = rand() * 0.35 + 0.55
        trns = 0.01
        mask = 0.5 .* (tanh.((xc .- xL_m) ./ trns) .- tanh.((xc .- xR_m) ./ trns))
        u .*= mask
    end

    u .-= minimum(u)
    if maximum(u) > 0
        u ./= maximum(u)
    end
    return u
end

# ── Dataset generator ─────────────────────────────────────────────────────────

"""
    run_parallel(root; N_ic=80, N_bc=80, seed=42, filename="RD_neumann_train")

Generate a reaction-diffusion dataset with `N_ic` initial conditions and
`N_bc` boundary condition pairs, and save to an HDF5 file.

# Arguments
  - `root`: Directory in which to save the HDF5 file
  - `N_ic`: Number of initial conditions
  - `N_bc`: Number of Neumann BC pairs (gL, gR)
  - `seed`: Random seed
  - `filename`: Base filename (without extension)
"""
function run_parallel(root; N_ic=80, N_bc=80, seed=42, filename="RD_neumann_train")
    Random.seed!(seed)
    ic_array = [generate_ic(x_RD) for _ in 1:N_ic]
    bc_array = [(round(0.05 * rand(), digits=3), round(-0.05 * rand(), digits=3))
                for _ in 1:N_bc]

    mkpath(root)
    path = joinpath(root, "$(filename)_nIC$(N_ic)_nBC$(N_bc).h5")

    h5open(path, "w") do f
        # Metadata
        f["ic"] = Float32.(hcat(ic_array...))  # (Nx, N_ic)
        bc_mat  = Float32.(hcat([[bc[1], bc[2]] for bc in bc_array]...))  # (2, N_bc)
        f["bc"] = bc_mat
        f["x"]  = Float32.(x_RD)
        f["t"]  = Float32.(t_grid_RD)
        HDF5.attributes(f)["rho"] = rho_RD
        HDF5.attributes(f)["nu"]  = nu_RD

        # Python/h5py shape: (N_ic, N_bc, Nx, nt)
        # Julia HDF5 reversed dims: (nt, Nx, N_bc, N_ic)
        d_create(f, "u", datatype(Float32),
                 dataspace(nt_RD, Nx_RD, N_bc, N_ic))

        Threads.@threads for i_ic in 1:N_ic
            for i_bc in 1:N_bc
                gL, gR = bc_array[i_bc]
                sol = solve_single(ic_array[i_ic], gL, gR)  # (Nx, nt)
                # Store as (nt, Nx) so Python reads (Nx, nt)
                f["u"][:, :, i_bc, i_ic] = Float32.(permutedims(sol, (2, 1)))
            end
        end
    end

    println("Saved to $path")
    return path
end

# ── Main script ───────────────────────────────────────────────────────────────

if abspath(PROGRAM_FILE) == @__FILE__
    # Training data: vary IC and BC
    run_parallel("datasets/data/"; N_ic=80, N_bc=80, seed=42, filename="RD_neumann_train")
    run_parallel("datasets/data/"; N_ic=30, N_bc=30, seed=0,  filename="RD_neumann_test")

    # Sampling data: fixed ICs, many BCs
    run_parallel("datasets/data/"; N_ic=20, N_bc=512, seed=42, filename="RD_sampling_diffICs")
end

File: archive/datasets_py/generate_RD1d_data.py
# script to solve the Reaction-Diffusion equation numerically to construct the PDE solution datasets for training and sampling

import numpy as np
import h5py
import os
import multiprocessing as mp
from functools import partial

rho, nu = 0.01, 0.005
Nx = 128
xL, xR = 0.0, 1.0
dt_save = 0.01
ini_time, fin_time = 0.0, 1.0 - dt_save
CFL = 0.25

x = np.linspace(xL, xR, Nx, endpoint=False) + (xR - xL) / Nx * 0.5
dx = x[1] - x[0]
nt = int(np.ceil((fin_time - ini_time) / dt_save)) + 1
t_grid = np.linspace(ini_time, fin_time, nt)

def solve_single(u0, gL, gR):
    u = u0.copy()
    sol = np.zeros((nt, Nx))
    sol[0] = u
    t = ini_time
    save_idx = 1

    def Courant_diff(dx, epsilon=1.0e-3):
        return 0.5 * dx**2 / (epsilon + 1e-8)

    def flux_numpy(u):
        N = u.size
        _u = np.zeros(N + 4)
        _u[2:2 + N] = u
        f = -nu * (_u[2:N + 3] - _u[1:N + 2]) / dx
        f[0] = gL
        f[-1] = gR
        return f

    def update(u, u_tmp, dt):
        stiff = 1.0 / (1.0 + np.exp(-rho * dt) * (1.0 - u) / (u + 1e-12))
        f = flux_numpy(u_tmp)
        return stiff - dt * (f[1:u.size + 1] - f[0:u.size]) / dx

    while t < fin_time and save_idx < nt:
        dt = min(Courant_diff(dx, nu) * CFL, fin_time - t, t_grid[save_idx] - t)
        if dt <= 1e-8:
            break
        u_tmp = update(u, u, 0.5 * dt)
        u = update(u, u_tmp, dt)
        t += dt
        if t >= t_grid[save_idx] - 1e-12:
            sol[save_idx] = u
            save_idx += 1
    sol[-1] = u
    return sol

def generate_ic(xc, k_tot=3, num_choice_k=2):
    selected = np.random.choice(k_tot, size=(num_choice_k,), replace=True)
    onehot = np.zeros(k_tot, dtype=int)
    for j in selected:
        onehot[j] += 1
    kk = 2 * np.pi * np.arange(1, k_tot + 1) * onehot / (xc[-1] - xc[0])
    amp = np.random.rand(k_tot, 1)
    phs = 2 * np.pi * np.random.rand(k_tot, 1)
    u = (amp * np.sin(kk[:, None] * xc[None, :] + phs)).sum(axis=0)
    if np.random.rand() < 0.1:
        u = np.abs(u)
    u *= np.random.choice([1, -1])
    if np.random.rand() < 0.1:
        xL = np.random.uniform(0.1, 0.45)
        xR = np.random.uniform(0.55, 0.9)
        trns = 0.01
        mask = 0.5 * (np.tanh((xc - xL) / trns) - np.tanh((xc - xR) / trns))
        u *= mask
    u -= u.min()
    if u.max() > 0:
        u /= u.max()
    return u

def worker(args):
    i_ic, i_bc, ic_array, bc_array = args
    u0 = ic_array[i_ic]
    gL, gR = bc_array[i_bc]
    sol = solve_single(u0, gL, gR)
    print(f"Finished: IC #{i_ic}, BC #{i_bc}, gL={gL:.3f}, gR={gR:.3f}")
    return i_ic, i_bc, sol.T

def run_parallel(root, N_ic=80, N_bc=80, nproc=48, seed=42, filename="RD_neumann_train"):
    np.random.seed(seed)
    ic_array = [generate_ic(x) for _ in range(N_ic)]
    bc_array = [(np.round(0.05 * np.random.rand(), 3), np.round(-0.05 * np.random.rand(), 3)) for _ in range(N_bc)]

    os.makedirs(root, exist_ok=True)
    path = os.path.join(root, f'{filename}_nIC{N_ic}_nBC{N_bc}.h5')

    with h5py.File(path, 'w') as f:
        f.create_dataset('u', shape=(N_ic, N_bc, Nx, nt), dtype=np.float32)
        f.create_dataset('ic', data=np.stack(ic_array), dtype=np.float32)
        f.create_dataset('bc', data=np.array(bc_array), dtype=np.float32)
        f.create_dataset('x', data=x, dtype=np.float32)
        f.create_dataset('t', data=t_grid, dtype=np.float32)
        f.attrs['rho'] = rho
        f.attrs['nu'] = nu

        tasks = [(i, j, ic_array, bc_array) for i in range(N_ic) for j in range(N_bc)]
        with mp.Pool(processes=nproc) as pool:
            for i_ic, i_bc, sol in pool.imap_unordered(worker, tasks):
                f['u'][i_ic, i_bc] = sol.astype(np.float32)
    return path


# training data: vary IC and BC 
run_parallel(root='datasets/data/', N_ic=80, N_bc=80, nproc=48, seed=42, filename="RD_neumann_train")
run_parallel(root='datasets/data/', N_ic=30, N_bc=30, nproc=48, seed=0, filename="RD_neumann_test")

# sampling data for fixed ICs
run_parallel(root='datasets/data/', N_ic=20, N_bc=512, nproc=48, seed=42, filename="RD_sampling_diffICs")
